torch/lobpcg.py

```diff
@@ -0,0 +1,562 @@
+              tensor U will be empty.
+
+    Arguments:
+
+      M (Tensor) : positive definite matrix, size is (m, m)
+      U (Tensor) : initial approximation, size is (m, n)
+      V (Tensor) : M-orthogonal external basis, size is (m, k)
+      tau_ortho (float, optional) : positive tolerance for orthogonality checks.
+      tau_drop (float, optional) : positive tolerance passed to svqb when called in drop mode
+      tau_replace (float, optional) : positive tolerance passed to svqb when called in replace mode
+      use_drop (bool, optional) : when True, enable dropping U columns
+                   that have small contribution to the `span([U, V])`.
+
+    Returns:
+
+      U (Tensor) : M-orthonormal columns (:math:`U^T M U = I`) such
+                   that :math:`V^T M U=0`, size is (m, n1), where `n1
+                   = n` if `drop` is `False, otherwise `n1 <= n`.
+
+      stats (dict) : statistics information
+    """
+    mm = torch.matmul
+    mm_M = _utils.get_matmul(M)
+    MV_norm = _utils.norm(mm_M(M, V))
+    MU = mm_M(M, U)
+    VMU = mm(_utils.transpose(V), MU)
+    i = j = 0
+    i_max = params.get('ortho.i_max', 3)
+    j_max = params.get('ortho.j_max', 3)
+    stats = {'ortho.UMUmI_rerr': {},
+             'ortho.VMU_rerr': {}}
+    for i in range(i_max):
+        U = U - mm(V, VMU)
+        drop = False
+        tau_svqb = tau_drop
+        for j in range(j_max):
+            if use_drop:
+                U = svqb(M, U, drop=drop, tau=tau_svqb)
+                drop = True
+                tau_svqb = tau_replace
+            else:
+                U = svqb(M, U, drop=False, tau=tau_replace)
+            if torch.numel(U) == 0:
+                # all initial U columns are M-collinear to V
+                stats['ortho.i'] = i
+                stats['ortho.j'] = j
+                return U, stats
+            MU = mm_M(M, U)
+            UMU = mm(_utils.transpose(U), MU)
+            U_norm = _utils.norm(U)
+            MU_norm = _utils.norm(MU)
+            R = UMU - torch.eye(UMU.shape[-1],
+                                device=UMU.device,
+                                dtype=UMU.dtype)
+            R_norm = _utils.norm(R)
+            rerr = R_norm / (MU_norm * U_norm)
+            if rerr < tau_ortho:
+                stats['ortho.UMUmI_rerr'][i, j] = rerr
+                break
+        VMU = mm(_utils.transpose(V), MU)
+        VMU_norm = _utils.norm(VMU)
+        rerr = VMU_norm / (MV_norm * U_norm)
+        if rerr < tau_ortho:
+            stats['ortho.VMU_rerr'][j] = rerr
+            break
+        if M is not None and M.shape[-1] < U.shape[-1] + V.shape[-1]:
+            raise ValueError(
+                'Overdetermined shape of U:'
+                ' #M-cols(={}) >= #U-cols(={}) + #V-cols(={}) must hold'
+                .format(M.shape[-1], U.shape[-1], V.shape[-1]))
+    stats['ortho.i'] = i
+    stats['ortho.j'] = j
+    return U, stats
```

Comment:
Yes, it is private. It would be good to keep the name to contain "ortho" to follow the convention used in several papers that describe this algorithm.